keyboards/duck/tcv3/matrix.c

```diff
@@ -0,0 +1,280 @@
+/*
+Copyright 2019 MechMerlin <mechmerlin@gmail.com>
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <util/delay.h>
+#include <avr/io.h>
+#include <stdio.h>
+#include "matrix.h"
+#include "util.h"
+#include "print.h"
+#include "debug.h"
+
+static uint8_t debouncing = DEBOUNCE;
+
+/* matrix state(1:on, 0:off) */
+static matrix_row_t matrix[MATRIX_ROWS];
+static matrix_row_t matrix_debouncing[MATRIX_ROWS];
+
+static uint8_t read_rows(uint8_t col);
+static void init_rows(void);
+static void unselect_cols(void);
+static void select_col(uint8_t col);
+
+
+__attribute__ ((weak))
+void matrix_init_kb(void) {
+    matrix_init_user();
+}
+
+__attribute__ ((weak))
+void matrix_scan_kb(void) {
+    matrix_scan_user();
+}
+
+__attribute__ ((weak))
+void matrix_init_user(void) {
+}
+
+__attribute__ ((weak))
+void matrix_scan_user(void) {
+}
+
+void backlight_init_ports(void)
+{
+  // DDRD  |=  0b11010000;
+  // PORTD &= ~0b01010000;
+  // PORTD |=  0b10000000;
+  // DDRB  |=  0b00011111;
+  // PORTB &= ~0b00001110;
+  // PORTB |=  0b00010001;
+  // DDRE  |=  0b01000000;
+  // PORTE &= ~0b01000000;
+}
+
+void matrix_init(void) {
+  backlight_init_ports();
+  unselect_cols();
+  init_rows();
+
+  for (uint8_t i=0; i < MATRIX_ROWS; i++)  {
+    matrix[i] = 0;
+    matrix_debouncing[i] = 0;
+  }
+
+  matrix_init_quantum();
+}
+
+uint8_t matrix_scan(void) {
+  for (uint8_t col = 0; col < MATRIX_COLS; col++) {
+    select_col(col);
+    _delay_us(3);
+
+    uint8_t rows = read_rows(col);
+
+    for (uint8_t row = 0; row < MATRIX_ROWS; row++) {
+      bool prev_bit = matrix_debouncing[row] & ((matrix_row_t)1<<col);
+      bool curr_bit = rows & (1<<row);
+      if (prev_bit != curr_bit) {
+        matrix_debouncing[row] ^= ((matrix_row_t)1<<col);
+        if (debouncing) {
+            dprint("bounce!: "); dprintf("%02X", debouncing); dprintln();
+        }
+        debouncing = DEBOUNCE;
+      }
+    }
+    unselect_cols();
+  }
+
+  if (debouncing) {
+    if (--debouncing) {
+      _delay_ms(1);
+    } else {
+      for (uint8_t i = 0; i < MATRIX_ROWS; i++) {
+        matrix[i] = matrix_debouncing[i];
+      }
+    }
+  }
+
+  matrix_scan_quantum();
+  return 1;
```

Comment:
Always changed?

Reply:
what do you mean?

Reply:
it should return whether the matrix changed since last scan

Reply:
returning 1 means that the matrix has changed.  Returning 0 means that it hasn't. 

So returning 1 means that it has always changed.  This doesn't affect a lot of things, since not as much relies on it. But OLED Driver definitely does.

Reply:
```suggestion
  return (uint8_t)changed;
```